internals/idkey/idkey_test.go

```diff
@@ -0,0 +1,154 @@
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package idkey_test
+
+import (
+	"crypto"
+	"crypto/ed25519"
+	"crypto/rand"
+	"os"
+	"path/filepath"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/canonical/pebble/internals/idkey"
+)
+
+// TestNoDirectory checks if leaf directory creation works.
+func (ks *keySuite) TestNoDirectory(c *C) {
+	keyDir := filepath.Join(c.MkDir(), "identity")
+
+	// Create a new identity key (first boot)
+	firstBoot, err := idkey.GenerateKey(keyDir)
+	c.Assert(err, IsNil)
+
+	// Load the identity key (other boots)
+	nextBoot, err := idkey.LoadKey(keyDir)
+	c.Assert(err, IsNil)
+
+	// Both should be the same identity.
+	c.Assert(firstBoot.Fingerprint(), Equals, nextBoot.Fingerprint())
+}
+
+// TestNew checks if the New() function correctly only creates a new
+// identity the first time.
+func (ks *keySuite) TestNew(c *C) {
+	keyDir := filepath.Join(c.MkDir(), "identity")
+
+	// Create a new identity key (first boot)
+	firstBoot, err := idkey.New(keyDir)
+	c.Assert(err, IsNil)
+
+	// Load the identity key (other boots)
+	nextBoot, err := idkey.New(keyDir)
+	c.Assert(err, IsNil)
+
+	// Both should be the same identity.
+	c.Assert(firstBoot.Fingerprint(), Equals, nextBoot.Fingerprint())
+}
+
+// TestDirectoryInvalid confirms that if the leaf directory has invalid
+// permissions we exit.
+func (ks *keySuite) TestDirectoryInvalid(c *C) {
+	keyDir := filepath.Join(c.MkDir(), "identity")
+	err := os.MkdirAll(keyDir, 0o740)
+	c.Assert(err, IsNil)
+
+	// Loading
+	_, err = idkey.LoadKey(keyDir)
+	c.Assert(err, ErrorMatches, ".* expected permission 0o700 .*")
+
+	// Saving
+	_, err = idkey.GenerateKey(keyDir)
+	c.Assert(err, ErrorMatches, ".* expected permission 0o700 .*")
+}
+
+// TestDirInvalid checks for a missing non-leaf directory. The caller
+// must create this.
+func (ks *keySuite) TestDirInvalid(c *C) {
+	keyDir := filepath.Join(c.MkDir(), "foo/identity")
+
+	// Saving
+	_, err := idkey.GenerateKey(keyDir)
+	c.Assert(err, ErrorMatches, "cannot create identity directory.*")
+}
+
+// TestInvalidKey checks permission of the key file.
+func (ks *keySuite) TestInvalidKey(c *C) {
+	keyDir := filepath.Join(c.MkDir(), "identity")
+
+	// Create a new identity key (first boot)
+	_, err := idkey.GenerateKey(keyDir)
+	c.Assert(err, IsNil)
+
+	err = os.Chmod(filepath.Join(keyDir, "key.pem"), 0o644)
+	c.Assert(err, IsNil)
+
+	// Load the identity key (other boots)
+	_, err = idkey.LoadKey(keyDir)
+	c.Assert(err, ErrorMatches, "cannot load identity key.*")
```

Comment:
I think we should have a more specific check here:
```suggestion
	c.Assert(err, ErrorMatches, ".*expected permissions.*")
```

Reply:
Fixed, thanks. Resolving.